### DISCUSSION TITLE
TextPage creation for other OCR engine

### DISCUSSION THREAD

**Discussion by darwinharianto** [pymupdf/PyMuPDF#2299] [Looking for help] [answered]
I have tried tesseract-ocr-jpn to create textpage, but since the result isn't as good as I expected, is it possible to swap out to other engines?

I wanted to use easyOCR as my textpage back-end, but I am not sure how to construct the TextPage class itself.

What kind of data or structure do I need to prepare for constructing TextPage object?

**Comment by JorjMcKie** [comment, marked as answer]
This is not supported, sorry.
have you tried to increase the `dpi` value when using `page.get_textpage_ocr()`?

**Reply by darwinharianto** [reply]
Yes, I tried from 72, 96, 256, up to 1000. I find that it is missing some part of the text.
Is some preprocessing being done inside that get_textpage_ocr method?

**Reply by darwinharianto** [reply]
Can I construct TextPage object from python?
edit: I am thinking to dump the ocr results from other object, and convert them to pymupdf's TextPage, so I can use the methods inside it

**Reply by JorjMcKie** [reply]
> Can I construct TextPage object from python? edit: I am thinking to dump the ocr results from other object, and convert them to pymupdf's TextPage, so I can use the methods inside it

This is indeed possible - however with an undocumented page method `page.extend_textpage()`. So you would need a little of experimenting.
The method is already used inside `page.get_textpage_ocr(..., full=False)`: There, images on a given page are separately OCR-ed (each producing a little temporary PDF page having its own respective TextPage) and the respective result is being merged into the "normal" TextPage.
You can try and use a similar approach.
When you use a different OCR package, it should produce a PDF page (let's call it "temp_page") with a text layer produced by it. Make a TextPage from that page and extend the TextPage of your original page with that TextPage.
So the process could look like this:
1. Make TextPage of your page: `tp = page.get_textpage()`
2. OCR the page with your other product. The result should be another PDF with its respective page, which we call "temp_page".
3. Make a TextPage from temp_page, but extend the original textpage tp with it:

```python
temp_page.extend_textpage(tp, flags=flags, matrix=mat)
```

4. Thereafter, you can do things like `page.get_text(option, textpage=tp)` where "option" is one of the usual strings "dict", "text", etc.

The matrix "mat" above transforms the "temp_page" rectangle to the "page" rectangle, so should probably just be the identity matrix. Otherwise, you can make a matrix that transforms the temp_page rectangle to the page rectangle like this: `mat = temp_page.rect.torect(page.rect)`.

**Reply by darwinharianto** [reply]
Thank you!
I plan to create the textpage fully using OCR results.
I am not sure on the "temp_page" part. It seems to be created from 
```            
imgdoc = Document(
                    "pdf", pix.pdfocr_tobytes(language=language)
                 )  # pdf with OCRed page
imgpage = imgdoc.load_page(0)  # read image as a page
```
How do I create that `pix.pdfocr_tobytes(language=language)`?

OCR result seems to be outputting bboxes with text inside it
So is it correct with this approach?

1. create empty page and shape
2. use `shape.insert_text`  to fill the pdf, while calculating the font size, rotation, else using the results from ocr

will this fill all the other details?
such as blockno or lineno?

**Reply by JorjMcKie** [reply]
I should have mentioned that TextPage has an independent constructor: `tp = fitz.TextPage(rect)`. Best use `fitz.INFINITE_RECT()` for rect.

But otherwise your suggestion summarized under points 1. & 2. is correct (and you can ignore my comments on TextPage altogether).
Just be sure you use everything provided by your OCR machine: fontsize, font (!), etc. Calculation of the insertion point of each text piece is especially interesting:
Even if e.g. the words "hello" and "you" have the same font size in the original and occur on the same line, the bottoms of their bboxes are different, because of the "y" in "you". If you insert the words, the baseline of "you" will be above that of "hello", because text insertion will use the bottom-left of the bbox in both cases.
Either you use the font provided by the OCR and choose an insertion point somewhat higher in cases where the word contains descending characters (take font.descender for computing this) or try using a common descender like -0.2:
For the word bbox `bbox.bl + (0, descender * bbox.height)` as insertion point if there are characters lapping below the baseline: y, j, |, etc. Note that descender values are negative.

The fontsize to choose is a similar thing: you should just use bbox.height for this, because in the original "Hello" and "you" may be the same fontsize - there bboxes will be different nonetheless. Unfortunately, there is no easy solution here. So best is to compute the fontsize by taking bbox.width into account: in `font.text_length(text, fontsize=fs)` the result must equal bbox.width.

**Reply by darwinharianto** [reply]
I ended up using fitz.Font(fontname="japan") as the font and calculating the size using font size using 
```
def calculate_font_size(boxWidth:float, text:str, font):
        for i in range(1000):
            length = font.text_length(text, fontsize=i)
            if length > boxWidth:
                return i-1
```
for japanese text the size seems to be calculated correctly, but for alphabet and numbers it seems to be stretched
Looks like it become full width for alphabets and numbers instead of being written as `123` it is written as `１２３`.
Is there a way to prevent it being written as a full width? 
or maybe make the text_length to calculate it using full width size?

Edit:
Just realized that somehow the page becomes an orphaned....
Not sure why

```
oldPage = doc[0]


page = doc.new_page(width, height)
# fill page with data from ocr

page.extend_textpage(oldPage.get_textpage(), flags=0, matrix=fitz.IdentityMatrix())
textPage = oldPage.get_textpage()
```

> oldPage.get_textpage()

this part raises ValueError: orphaned object: parent is None

**Reply by JorjMcKie** [reply]
If using `page.insert_text*()` methods with CJK fonts keywords ("japan", "korea", ...) lead to using non-embedded fonts. And those have the unpleasant peculiarity to be monospaced with the effect you have observed: Latin and numbers are awkwardly positioned within the character space.
I recommend using the buffer of font `fitz.Font("cjk")` instead:
```python
fontbuffer = fitz.Font("cjk").buffer
# then 'install' that font on any page with Asian text
page.insert_font(fontname="cjk", fontbuffer=fontbuffer)
...
page.insert_text(point,txt, ..., fontname="cjk")
```
This should lead to a nice look. The CJK font will be embedded however, and is quite large (3.6 MB uncompressed). So make sure you (1) have package fontTools installed, and (2) invoke `doc.subset_fonts()` before saving.

The "orphan" problem: Do not dynamically create the to-be-extended text page within method `page.extend_textpage()` because it will be deleted instantly again. Instead create it before `.extend_textpage()` and keep it alive all the way down ...

**Reply by JorjMcKie** [reply]
If you need to compute text lengths, you should keep the CJK font "alive" via e.g. `cjk = fitz.Font("cjk")`. then use `cjk.buffer` when you need its buffer for using `.insert_text()`. Text length computations work as expected using `cjk.text_length(...)`.
That font supports Chinese, Japanese, Korean, Latin including Cyrillic and Greece.

**Reply by darwinharianto** [reply]
Thanks you! 
cjk.text_length(...) really helps with the character spacing!

I plan to show the OCR confidence using opacity. They are perfectly shown using inset_text(xxx_opacity= confidence), but how can I extract the opacity data?
Is it only available inside `page.get_texttrace()`?
can I get it from textpage? 
I can't find opacity key inside `textpage.extractRawDict`

**Reply by JorjMcKie** [reply]
>  Is it only available inside page.get_texttrace()?

Yes, unfortunately this is one of the properties not available in `TextPage`, which is a MuPDF object and the data source for all `page.get_text()` and `page.search_for()` methods and variants. Although that may change in the future, there are no plans to do so to my best knowledge.

`page.get_texttrace()` is **not** based on TextPage, but directly walks through the page appearance instructions. Therefore, it can potentially return all properties of text like opacity (for both, fill and stroke operations) and text rendering mode. On the other hand it does lack the advanced text preprocessing features of TextPage, like the block-line-span-char hierarchy, dynamic space character generation and more.

**Reply by darwinharianto** [reply]
As a closing remarks...
to create a new textpage based on the ocr-ed page
```

def calculate_font_size(rect:Rect, text:str, font, rotation:float):
    # there are other faster methodd
    for i in range(1000):
        length = font.text_length(text, fontsize=i)
        if length > rect.width and (rotation == 0 or rotation == 180):
            return i-1
        elif length > rect.height and (rotation == 90 or rotation == 270):
            return i-1
    return I

newPage = doc.new_page(width=page.rect.width, height=page.rect.height)
page.set_rotation(page.rotation)
page.insert_font(fontname="cjk", fontbuffer=self.font.buffer)
shape = page.new_shape()
font = fitz.Font(fontname="cjk")

# this OCR results is from easyocr
for ocrResult in ocrResults:
      rect = fitz.Rect(ocrResult[0][0] + ocrResult[0][2])
      text = ocrResult[1]
      confidence = ocrResult[2]
      fontsize = OCRPDFProc.calculate_font_size(rect, text, font, rotation=0)
      # we can represent confidence with opacity
      shape.insert_text(rect.bl, text, fontsize, fontname='cjk', rotate=0, stroke_opacity=confidence, fill_opacity=confidence)
shape.commit()
```

**Reply by hiroi-sora** [reply]
Thank you very much!

I am the author of [Umi-OCR](https://github.com/hiroi-sora/Umi-OCR), currently developing a PDF recognition feature to convert scanned documents into searchable PDFs. My project also uses other OCR engines, such as [PaddleOCR-json](https://github.com/hiroi-sora/PaddleOCR-json).

Your idea has provided me with great inspiration; the positioning of the generated invisible text blocks is very accurate. Below is my improved `calculate_font_size` function, which can reduce the number of searches:

```python
    # Calculate the font size that fills the width and height for a line of text
    def _calculateFontSize(self, text, w, h):
        if h > w:  # For vertical text, switch to horizontal for calculation
            w, h = h, w
        fontsize = round(h)  # Initial font size, assumed to be line height
        minSize = 5  # Minimum size limit
        getLen = lambda text, s: self.font.text_length(text, fontsize=s)
        while getLen(text, fontsize) < w:
            fontsize += 1  # Try increasing the font size until the line width just exceeds the limit
        fontsize -= 1
        while getLen(text, fontsize) > w and fontsize >= minSize:
            fontsize -= 1  # Try decreasing the font size until the line width is just less than the limit
        return fontsize
```

**Reply by nissansz** [reply]
> Thanks you! cjk.text_length(...) really helps with the character spacing!
> 
> I plan to show the OCR confidence using opacity. They are perfectly shown using inset_text(xxx_opacity= confidence), but how can I extract the opacity data? Is it only available inside `page.get_texttrace()`? can I get it from textpage? I can't find opacity key inside `textpage.extractRawDict`

实际显示的效果是怎么样的？发个图片看看
